common/djangoapps/student/tests/test_models.py

```diff
@@ -403,3 +403,112 @@ def test_retire_recovery_email(self):
+        actual_student = User.objects.get(email=student.email)
+        actual_cea = CourseEnrollmentAllowed.objects.get(email=student.email)
+
+        self.assertEqual(actual_course_enrollment.mode, mode)
+        self.assertEqual(actual_student.is_active, True)
+        self.assertEqual(actual_cea.user, student)
+
+    def test_not_enrolled_student_is_enrolled(self):
+        """
+        Make sure that invited students who are not enrolled become enrolled when their account is activated.
+        They should be enrolled in the course in audit mode.
+        """
+        # fixture
+        student = self._set_up_invited_student(
+            course=self.course,
+            active=False,
+            enrolled=False
+        )
+
+        # trigger the post_save callback
+        student.is_active = True
+        student.save()
+
+        # reload values from the database + make sure they are in the expected state
+        actual_course_enrollment = CourseEnrollment.objects.get(user=student, course_id=self.course.id)
+        actual_student = User.objects.get(email=student.email)
+        actual_cea = CourseEnrollmentAllowed.objects.get(email=student.email)
+
+        self.assertEqual(actual_course_enrollment.mode, u"audit")
+        self.assertEqual(actual_student.is_active, True)
+        self.assertEqual(actual_cea.user, student)
+
+    def test_verified_student_not_downgraded_when_changing_email(self):
+        """
+        Make sure that verified students do not get downgrade if they are active + changing their email.
+        """
+        # fixture
+        student = self._set_up_invited_student(
+            course=self.course,
+            active=True,
+            course_mode=u'verified'
+        )
+        old_email = student.email
+
+         # trigger the post_save callback
+        student.email = "foobar" + old_email
+        student.save()
+
+        # reload values from the database + make sure they are in the expected state
+        actual_course_enrollment = CourseEnrollment.objects.get(user=student, course_id=self.course.id)
+        actual_student = User.objects.get(email=student.email)
+
+        self.assertEqual(actual_course_enrollment.mode, u"verified")
+        self.assertEqual(actual_student.is_active, True)
+
+    def _set_up_invited_student(self, course, active=False, enrolled=True, course_mode=''):
+        """
+        Helper function to create a user in the right state, invite them into the course, and update their
+        course mode if needed.
+        """
+        email = 'robot@robot.org'
+        user = UserFactory(
+            username='somestudent',
+            first_name='Student',
+            last_name='Person',
+            email=email,
+            is_active=active
+        )
+
+        # invite the user to the course
+        cea = CourseEnrollmentAllowed(email=email, course_id=course.id, auto_enroll=True)
+        cea.save()
+
+        if enrolled:
```

Comment:
Not super enthused about adding logic into a test helper but also didn't want to have the unit tests do any of the fixture setup...